components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java

```diff
@@ -4726,6 +4742,54 @@ public API getAPIbyUUID(String uuid, String organization) throws APIManagementEx
         }
     }
 
+    @Override
+    public APISearchResult searchPaginatedAPIsByFQDN(String endpoint, String tenantDomain,
+                                                     int offset, int limit) throws APIManagementException {
+        String fqdn;
+        APISearchResult result = new APISearchResult();
+
+        try {
+            URI uri = new URI(endpoint);
+            fqdn = uri.getHost();
+            if(fqdn == null) {
+                return result;
+            }
+        } catch (URISyntaxException e) {
+            throw new APIManagementException("Error while extracting fully qualified domain name from the given url: " + endpoint, e);
+        }
+
+        String query = ENDPOINT_CONFIG_SEARCH_TYPE_PREFIX + fqdn;
+        Organization org = new Organization(tenantDomain);
+        String adminUser = APIUtil.getTenantAdminUserName(tenantDomain);
+        String[] roles = APIUtil.getFilteredUserRoles(adminUser);
+        Map<String, Object> properties = APIUtil.getUserProperties(adminUser);
+        UserContext userCtx = new UserContext(adminUser, org, properties, roles);
+
+        try {
+            PublisherAPISearchResult searchAPIs = apiPersistenceInstance.searchAPIsForPublisher(org, query,
+                    offset, limit, userCtx, "createdTime", "desc");
+            if (log.isDebugEnabled()) {
+                log.debug("Running Solr query : " + query);
+            }
+            if (searchAPIs != null) {
+                List<PublisherAPIInfo> list = searchAPIs.getPublisherAPIInfoList();
+                List<API> apiList = new ArrayList<>();
+                for (PublisherAPIInfo publisherAPIInfo : list) {
+                    API mappedAPI = APIMapper.INSTANCE.toApi(publisherAPIInfo);
+                    populateAPIStatus(mappedAPI);
+                    populateDefaultVersion(mappedAPI);
+                    apiList.add(mappedAPI);
+                }
+                result.setApis(apiList);
+                result.setApiCount(searchAPIs.getTotalAPIsCount());
```

Comment:
Shouldn't we set the size of the apiList? Does searchAPIs.getTotalAPIsCount() returns the same number as the size of searchAPIs.getPublisherAPIInfoList() in all the cases?

Reply:
1. Added an initial capacity to the apiList object
2. searchAPIs.getTotalAPIsCount() == total num of APIs that matched the query
searchAPIs.getPublisherAPIInfoList().size() ==  limit or ( total num of APIs - offset )